### PR TITLE
rgw/lc: Fix issues with non-current objects with instance empty

### PR DESCRIPTION
**Pull request by soumyakoduri**
When the bucket versioning is enabled, old plain object entry is converted to versioned by updating its instance as "null" in its raw head/old object. However its instance remains empty in the bi list entry. Same is the case for the entries created after versioning is suspended and re-enabled.

So to access such objects which are non-current, we need to set rgw_obj_key.instance as
1) "null" to read the actual raw obj and
2) empty while accessing/updating their bi entry.

Tests: https://github.com/ceph/s3-tests/pull/593/file

Fixes: https://tracker.ceph.com/issues/68402
Signed-off-by: Soumya Koduri <skoduri@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by soumyakoduri** [comment]
@mattbenjamin @cbodley @dang this seems day-1 issue. Please review and let me know if this is right approach to fix  this issue

Transition seem to be working fine and I am yet to run expiration tests.

File: src/rgw/driver/rados/rgw_rados.cc [hunk 1]
```diff
@@ -3394,6 +3394,14 @@ int RGWRados::Object::Write::write_meta(uint64_t size, uint64_t accounted_size,
   RGWBucketInfo& bucket_info = target->get_bucket_info();
 
   RGWRados::Bucket bop(target->get_store(), bucket_info);
+  if (bucket_info.versioned()) {
+    rgw_obj_key& obj_key = target->get_obj().key;
+    // instance is empty for bi entries created when bucket versioning
+    // was not enabled or suspended.
+    if (obj_key.instance == "null") {
+      obj_key.instance.clear();
```
[the remark this answers is not in the document]
Comment (soumyakoduri) [reply src/rgw/driver/rados/rgw_rados.cc:3402]: I see similar changes were done in RGWRados::Object::Delete::delete_obj() and RGWRados::set_attrs() but they are directly handling `index_op`. 

`RGWRados::transition_obj()` uses copy_obj_data() which uses processor. Is there any way to clear that instance just before processor.complete() in `copy_obj_data`.

**Comment by soumyakoduri** [comment]
> can we add a s3test case for the lifecycle policy in https://tracker.ceph.com/issues/68402?

Sure.

File: src/rgw/driver/rados/rgw_rados.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (cbodley) [reply src/rgw/driver/rados/rgw_rados.cc:3402]: i was hoping `transition_obj()` could clear the instance after calling `read_op.prepare()`, which is where `follow_olh()` looks up the current version instead

are you sure that `copy_obj_data()` does the wrong thing if we pass it an empty instance?

[the remark this answers is not in the document]
Comment (soumyakoduri) [reply src/rgw/driver/rados/rgw_rados.cc:3402]: you are right. `copy_obj_data()` works fine even with instance empty.  I will add test-case too and update.

**Review by cbodley — APPROVED** [review]
looks good but adding `do-not-merge` label until the TEMP commit for qa/rgw/s3tests-branch.yaml is removed

i usually add that change to the branch i push to ceph-ci for testing, but leave it out of the pull request branch

**Comment by soumyakoduri** [comment]
http://pulpito.front.sepia.ceph.com/soumyakoduri-2024-10-21_19:32:51-rgw:cloud-transition-wip-skoduri-cloud-restore-distro-default-smithi
http://pulpito.front.sepia.ceph.com/soumyakoduri-2024-10-21_19:29:28-rgw:lifecycle-wip-skoduri-cloud-restore-distro-default-smithi

**Comment by soumyakoduri** [comment]
http://pulpito.front.sepia.ceph.com/soumyakoduri-2024-10-22_08:40:03-rgw-wip-skoduri-lc-nullinstance-distro-default-smithi/

**Pull request merged**